File: 104. Maximum Depth of Binary Tree.cpp
//Solution1
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        int res = 0;
        if (!root) return res;
        res++;
        queue<TreeNode*> q;
        q.push(root);
        int cur = 1, next = 0;
        while (!q.empty()) {
            TreeNode* root = q.front();
            q.pop();
            cur--;
            if (root->left) {
                q.push(root->left);
                next++;
            }
            if (root->right) {
                q.push(root->right);
                next++;
            }
            if (cur == 0 && next) {
                cur = next;
                next = 0;
                res++;
            }
        }
        return res;
    }
};

//Solution2
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if (!root) return 0;
        return 1 + max(maxDepth(root->left), maxDepth(root->right));
    }
};

File: java/104. Maximum Depth of Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
// O(n) recursive
class Solution {
    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
    }
}

// O(n) iterative BFS
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int count = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            while (size-- > 0) {
                TreeNode node = queue.poll();
                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }
            count++;
        }
        return count;
    }
}

// O(n) DFS iterative
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        Stack<TreeNode> nodes = new Stack<>();
        Stack<Integer> depths = new Stack<>();
        nodes.push(root);
        depths.push(1);
        int maxDepth = 0;
        while (!nodes.isEmpty()) {
            TreeNode cur = nodes.pop();
            int curDepth = depths.pop();
            maxDepth = Math.max(maxDepth, curDepth);
            if (cur.left != null) {
                nodes.push(cur.left);
                depths.push(curDepth + 1);
            }
            if (cur.right != null) {
                nodes.push(cur.right);
                depths.push(curDepth + 1);
            }
        }
        return maxDepth;
    }
}
